[WIP]  LibC: Use a reentrant lock in all stdio FILE

cbsirb opened the pull request:
Multiple threads writing to the same FILE would hit the assert where the
buffer index was bigger than size. This just uses a very basic lock
(stolen from pthread_mutex functions).

awesomekling commented:
It seems totally arbitrary to only do this for `fputc`. We should use a lock to protect all the I/O operationg on a `FILE*`, no?

awesomekling commented:
`fflush` also uses it. Locking everything sounds good :)

cbsirb commented:
Ok, I implemented locking everywhere (hope I didn't miss anything).

Also implemented the `flockfile`, `funlockfile` and `ftrylockfile`, and everything is based on those (+ a small helper class).

Made the lock re-entrant so we don't have a lot of `_no_lock` functions (fflush for eg. should have had one if not)

awesomekling commented:
Closing this for lack of activity (it's been WIP for months). Feel free to reopen if/when progress is made :)
